src/databricks/labs/remorph/reconcile/query_builder.py

```diff
@@ -0,0 +1,153 @@
+        transformations_dict = self.table_conf.list_to_dict(Transformation, "column_name")
+        column_mapping_dict = self.table_conf.list_to_dict(ColumnMapping, "target_name")
+
+        transformation_rule_mapping = []
+        for column in columns:
+            if column in transformations_dict.keys():
+                transformation = self._get_layer_transform(transformations_dict, column, self.layer)
+            else:
+                column_data_type = schema.get(column).data_type
+                transformation = self._get_default_transformation(self.source, column_data_type).format(column)
+
+            if column in column_mapping_dict.keys():
+                column_alias = column_mapping_dict.get(column).source_name
+            else:
+                column_alias = column
+
+            transformation_rule_mapping.append(TransformRuleMapping(column, transformation, column_alias))
+
+        return transformation_rule_mapping
+
+    @staticmethod
+    def _get_default_transformation(data_source: str, data_type: str) -> str:
+        match data_source:
+            case "oracle":
+                return OracleDataSource.oracle_datatype_mapper.get(
+                    data_type, ColumnTransformationType.ORACLE_DEFAULT.value
+                )
+            case "snowflake":
+                return SnowflakeDataSource.snowflake_datatype_mapper.get(
+                    data_type, ColumnTransformationType.SNOWFLAKE_DEFAULT.value
+                )
+            case "databricks":
+                return DatabricksDataSource.databricks_datatype_mapper.get(
+                    data_type, ColumnTransformationType.DATABRICKS_DEFAULT.value
+                )
+            case _:
+                msg = f"Unsupported source type --> {data_source}"
+                raise ValueError(msg)
+
+    @staticmethod
+    def _get_layer_transform(transform_dict: dict[str, Transformation], column: str, layer: str) -> str:
+        return transform_dict.get(column).source if layer == "source" else transform_dict.get(column).target
+
+    @staticmethod
+    def _get_column_expr(func, column_transformations: list[TransformRuleMapping]):
+        return [func(transformation) for transformation in column_transformations]
+
+    @staticmethod
+    def _generate_hash_algorithm(source: str, column_expr: list[str]) -> str:
+        if source in {SourceType.DATABRICKS.value, SourceType.SNOWFLAKE.value}:
+            hash_expr = "concat(" + ", ".join(column_expr) + ")"
+        else:
+            hash_expr = " || ".join(column_expr)
+
+        return (Constants.hash_algorithm_mapping.get(source.lower()).get("source")).format(hash_expr)
+
+    @staticmethod
+    def _construct_hash_query(table_name: str, query_filter: str, hash_expr: str, key_column_expr: list[str]) -> str:
+        sql_query = StringIO()
+        sql_query.write(f"select {hash_expr} as {Constants.hash_column_name}, ")
```

Comment:
check if this is returning write sql, I believe this will introduce an additional comma.

Reply:
checked it works well with query with join columns.in case of only hash validation.This will have an impact.

Reply:
the comma at the end should be moved to line 144